quarkchain/cluster/master.py

```diff
@@ -902,61 +903,85 @@ def log_error_and_exit(msg):
                 Logger.info("Checking root block height: {}".format(rb.header.height))
-            prev_rb = self.root_state.db.get_root_block_by_hash(
-                rb.header.hash_prev_block
-            )
-            if prev_rb.header.height != rb.header.height - 1:
-                log_error_and_exit(
-                    "Root block height {} mismatches previous block".format(
-                        rb.header.height
-                    )
+            rb_list = []
+            for i in range(self.env.arguments.check_db_rblock_batch):
+                count += 1
+                if rb.header.height < max(check_db_rblock_to, 1):
+                    break
+                rb_list.append(rb)
+                prev_rb = self.root_state.db.get_root_block_by_hash(
+                    rb.header.hash_prev_block
                 )
+                if prev_rb.header.height != rb.header.height - 1:
+                    log_error_and_exit(
+                        "Root block height {} mismatches previous block".format(
+                            rb.header.height
+                        )
+                    )
 
-            if prev_rb.header.get_hash() != rb.header.hash_prev_block:
-                log_error_and_exit(
-                    "Root block height {} mismatches previous block hash".format(
-                        rb.header.height
+                if prev_rb.header.get_hash() != rb.header.hash_prev_block:
+                    log_error_and_exit(
+                        "Root block height {} mismatches previous block hash".format(
+                            rb.header.height
+                        )
                     )
-                )
+                rb = prev_rb
 
             future_list = []
-            for mheader in rb.minor_block_header_list:
-                conn = self.get_slave_connection(mheader.branch)
-                request = CheckMinorBlockRequest(mheader)
-                future_list.append(
-                    conn.write_rpc_request(ClusterOp.CHECK_MINOR_BLOCK_REQUEST, request)
-                )
+            header_list = []
+            for crb in rb_list:
+                header_list.extend(crb.minor_block_header_list)
+                for mheader in crb.minor_block_header_list:
+                    conn = self.get_slave_connection(mheader.branch)
+                    request = CheckMinorBlockRequest(mheader)
+                    future_list.append(
+                        conn.write_rpc_request(
+                            ClusterOp.CHECK_MINOR_BLOCK_REQUEST, request
+                        )
+                    )
 
-            try:
-                self.root_state.add_block(rb, write_db=False, skip_if_too_old=False)
-            except Exception as e:
-                Logger.log_exception()
-                log_error_and_exit(
-                    "Failed to check root block height {}".format(rb.header.height)
-                )
+            for crb in rb_list:
```

Comment:
ideally this part (line 959 - 968) can be done in parallel as waiting for slave responses (line 970 - 978). I would assume the latter takes more time because of the amount of minor blocks to handle in a slave